### PR TITLE
Moved auth_verify to server, returning 204 or 401 as per nginx standards

### PR DESCRIPTION
**Pull request by robvanoostenrijk**
With the current proxy mode, saml-auth-proxy cannot be used for only verification purposes.

This change follows the functionality seen on oauth2-proxy, exposing a verify only endpoint which either returns `204 Accepted` or `401 Forbidden`.

This should work for nginx, traefik and Caddy.

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -168,3 +174,22 @@ func setupHttpClient(idpCaFile string) (*http.Client, error) {
...
+		if session != nil {
+			w.WriteHeader(204)
+			return
+		}
+
+		if err == samlsp.ErrNoSession {
+			w.WriteHeader(401)
+			return
+		}
+
```
Comment (itzg) [inline server/server.go:193]: Perhaps log any other error that makes it here?

File: server/server.go [hunk 2]
```diff
@@ -110,6 +111,11 @@ func Start(ctx context.Context, logger *zap.Logger, cfg *Config) error {
 	}
 
 	app := http.HandlerFunc(proxy.handler)
+	if cfg.AuthVerify {
+		http.Handle(cfg.AuthVerifyPath, authVerify(middleware))
+	}
+
+	http.Handle("/saml/sign_in", http.HandlerFunc(middleware.HandleStartAuthFlow))
```
Comment (itzg) [inline server/server.go:118]: Can you point me to docs (or oauth2-proxy source code) that shows how this flow proceeds?

**Review by itzg — COMMENTED** [review]
Thanks! Just a small tweak and question.

File: server/server.go (hunk 2 above)
Comment (robvanoostenrijk) [reply server/server.go:118]: https://oauth2-proxy.github.io/oauth2-proxy/features/endpoints/

Traefik specific example:
https://github.com/oauth2-proxy/oauth2-proxy/blob/master/contrib/local-environment/traefik/dynamic.yaml

File: server/server.go (hunk 1 above)
Comment (robvanoostenrijk) [reply server/server.go:193]: The function is unwrapped from crewjam/saml, I added the existing middleware error callback now.

Comment (itzg) [reply server/server.go:193]: Now I see that is the only error they document returning there. So,  no further change needed?

Comment (robvanoostenrijk) [reply server/server.go:193]: I don't believe so, tried to keep the same behaviour as the middleware.

**Comment by itzg** [comment]
And FYI here also about merge conflict.

**Comment by robvanoostenrijk** [comment]
Merge conflicts resolved.

**Review by itzg — APPROVED** [review]

**Comment by itzg** [comment]
Included in https://github.com/itzg/saml-auth-proxy/releases/tag/1.15.0

**Pull request merged**